keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -162,10 +163,106 @@ TODO: complete this list
+appropriate storage limit for a cluster is out-of-scope for this document, once
+a etcd storage limit selected, suggested maximum limits for custom resources
+are:
+
+| etcd storage limit | Suggested Maximum Limit: scope=cluster |
+| --- | --- |
+| 4GB | 40000 |
+| 8GB | 80000 |
+
+These limits aim to keep custom resource storage usage to less than half of the
+total cluster storage capacity for custom resources of 50kb or less in size.
+
+**Custom Resources per Definition:**
+
+For each custom resource definition, the limit on the number of custom resources
+can be found by taking the (median) object size of the custom resource and finding
+the the matching row in this table:
+
+| Object size | Suggested Maximum Limit: scope=namespace (5s p99 SLO) | Suggested Maximum Limit: scope=cluster (30s p99 SLO) |
+| --- | --- | --- |
+| <=10kb | 1500 | 10000 |
+| (10kb - 25kb] | 600 | 4000 |
+| (25kb - 50kb] | 300 | 2000 |
+
+The cluster scope indicates the total number of custom resources for that
+definition allowed in the entire cluster.
+
+The namespace scope indicates the total number of custom resources for that
+definition allowed in any particular namespace. The cumulative count of the
+custom resource across all namespaces must not exceed the cluster limit.
+
+Since, in practice, custom resources scale farther without conversion webhooks
+within the SLI/SLOs (roughly 2x according to our scale tests), custom resource
+definition authors should be careful to adhere to these limits so that in the
+future a webhook converter may safely be added as part of a custom resource
+version upgrade.
+
+_Note: For custom resources of custom resource definitions using `scope: Namespaced`: the scope=namespace
+suggested maximum limit indicates how many custom resource objects may be in each namespace,
+and the scope=cluster suggested maximum limit indicates how many custom resource objects may
+be in the cluster total. For custom resources of custom resource definitions using `scope: Cluster`: only
+the scope=cluster suggested maximum limit applies._
+
+**Conversion Webhooks:**
+
+Conversion Webhook SLOs are defined from the perspective of the conversion
+webhook. It does not include any api-server serialization/deserialization for
+making the request to the webhook, but it does include network latency.
+
+Given that the performance and scalability of conversion webhooks are the
+responsibility of their author, Custom resource scale targets are applied only for
+conversion webhooks that are within the following latencies for the above suggested
+maximum limits.
+
+| scope | Expected conversion Webhook SLO: p99 latency |
+| --- | --- |
+| resource | 50ms |
+| namespace | 1 seconds |
+| cluster | 6 seconds |
+
+The above object size and suggested maximum limits in the Custom Resources per
+Definition table applies to these conversion webhook SLOs. For example, for a
+list request for 1500 custom resource objects that are 10k in size, the resource
```

Comment:
Is the table trying to say "return a single object in 50ms, up to 1500 10k objects in 1s, 10000 10k objects in 6 seconds"?

If so, that mathematically doesn't make much sense, the latter implies that a single object must be processed in .6ms?

Reply:
I've inlined the object count details in the webhook SLOs to make it easier to understand. The 50ms for a single object is higher to account for a request serving constant factor. I've added a note.